### PR TITLE
Visual testing API

### PR DESCRIPTION
**Pull request by nadvolod**
This is PR designed to iron out the visual API. Once we agree on it we can work on making it happen.
Here is a real world example of a current usage of Screener for visual testing https://github.com/saucelabs-training/demo-java/blob/master/java8/visual-e2e/visuale2e.junit4.examples/src/test/java/com/saucedemo/VisualCrossPlatformTests.java

### CODE REVIEW

File: java/src/main/java/com/saucelabs/saucebindings/visual/SauceVisualOptions.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
+package com.saucelabs.saucebindings.visual;
+
+import com.saucelabs.saucebindings.SauceEnvironmentVariablesNotSetException;
+import org.openqa.selenium.MutableCapabilities;
+
+public class SauceVisualOptions {
+    private final MutableCapabilities visualCapabilities;
```
Comment (nadvolod) [inline java/src/main/java/com/saucelabs/saucebindings/visual/SauceVisualOptions.java:7]: Codacy found an issue: [Perhaps 'visualCapabilities' could be replaced by a local variable.](https://app.codacy.com/manual/SauceLabs/sauce_bindings/pullRequest?prid=6239731)

**Comment by nadvolod** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 5
           

Complexity increasing per file
==============================
- java/src/test/java/com/saucelabs/saucebindings/SauceVisualSession.java  3
- java/src/test/java/com/saucelabs/saucebindings/integration/VisualE2ETest.java  1
- java/src/main/java/com/saucelabs/saucebindings/visual/SauceVisualOptions.java  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/SauceLabs/sauce_bindings/pullRequest?prid=6239731&bid=20327768)

File: java/src/test/java/com/saucelabs/saucebindings/SauceVisualSession.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.Map;
+
+public class SauceVisualSession {
+    private SauceVisualOptions visualOptions;
+    private RemoteWebDriver driver;
+
+    public SauceVisualSession() {
```
Comment (nadvolod) [inline java/src/test/java/com/saucelabs/saucebindings/SauceVisualSession.java:20]: Codacy found an issue: [Avoid unnecessary constructors - the compiler will generate these for you](https://app.codacy.com/manual/SauceLabs/sauce_bindings/pullRequest?prid=6239731)

File: java/src/test/java/com/saucelabs/saucebindings/SauceVisualSession.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
+package com.saucelabs.saucebindings;
+
+import com.saucelabs.saucebindings.visual.SauceVisualOptions;
+import lombok.Getter;
+import lombok.Setter;
+import lombok.var;
```
Comment (nadvolod) [inline java/src/test/java/com/saucelabs/saucebindings/SauceVisualSession.java:6]: Codacy found an issue: [Avoid unused imports such as 'lombok.var'](https://app.codacy.com/manual/SauceLabs/sauce_bindings/pullRequest?prid=6239731)

File: java/src/test/java/com/saucelabs/saucebindings/SauceVisualSession.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        return driver;
+    }
+
+    protected RemoteWebDriver createRemoteWebDriver(URL url, MutableCapabilities capabilities) {
+        return new RemoteWebDriver(url, capabilities);
+    }
+
+    @Getter
+    @Setter
+    private String hubUrl = "https://hub.screener.io/wd/hub";
```
Comment (nadvolod) [inline java/src/test/java/com/saucelabs/saucebindings/SauceVisualSession.java:41]: Codacy found an issue: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/manual/SauceLabs/sauce_bindings/pullRequest?prid=6239731)

File: java/src/test/java/com/saucelabs/saucebindings/SauceVisualSession.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+import org.openqa.selenium.JavascriptExecutor;
+import org.openqa.selenium.MutableCapabilities;
+import org.openqa.selenium.remote.RemoteWebDriver;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.Map;
+
+public class SauceVisualSession {
+    private SauceVisualOptions visualOptions;
```
Comment (nadvolod) [inline java/src/test/java/com/saucelabs/saucebindings/SauceVisualSession.java:17]: Codacy found an issue: [Perhaps 'visualOptions' could be replaced by a local variable.](https://app.codacy.com/manual/SauceLabs/sauce_bindings/pullRequest?prid=6239731)

**Comment by nadvolod** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 8
           

Complexity increasing per file
==============================
- java/src/main/java/com/saucelabs/saucebindings/visual/SauceVisualOptions.java  4
- java/src/test/java/com/saucelabs/saucebindings/SauceVisualSession.java  3
- java/src/test/java/com/saucelabs/saucebindings/integration/VisualE2ETest.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/SauceLabs/sauce_bindings/pullRequest?prid=6239731&bid=20327768)

File: java/src/main/java/com/saucelabs/saucebindings/visual/SauceVisualOptions.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
```
Comment (nadvolod) [inline java/src/main/java/com/saucelabs/saucebindings/visual/SauceVisualOptions.java:7]: Codacy found an issue: [Perhaps 'visualCapabilities' could be replaced by a local variable.](https://app.codacy.com/manual/SauceLabs/sauce_bindings/pullRequest?prid=6239731)

File: java/src/main/java/com/saucelabs/saucebindings/visual/SauceVisualOptions.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    private String visualProjectName;
```
Comment (nadvolod) [inline java/src/main/java/com/saucelabs/saucebindings/visual/SauceVisualOptions.java:8]: Codacy found an issue: [Avoid unused private fields such as 'visualProjectName'.](https://app.codacy.com/manual/SauceLabs/sauce_bindings/pullRequest?prid=6239731)

File: java/src/test/java/com/saucelabs/saucebindings/SauceVisualSession.java [hunk 8]
```diff
@@ -0,0 +1,70 @@
+package com.saucelabs.saucebindings;
+
+import com.saucelabs.saucebindings.visual.SauceVisualOptions;
+import lombok.Getter;
+import lombok.Setter;
+import lombok.var;
```
Comment (nadvolod) [inline java/src/test/java/com/saucelabs/saucebindings/SauceVisualSession.java:6]: Codacy found an issue: [Avoid unused imports such as 'lombok.var'](https://app.codacy.com/manual/SauceLabs/sauce_bindings/pullRequest?prid=6239731)

File: java/src/test/java/com/saucelabs/saucebindings/SauceVisualSession.java [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+        return driver;
+    }
+
+    protected RemoteWebDriver createRemoteWebDriver(URL url, MutableCapabilities capabilities) {
+        return new RemoteWebDriver(url, capabilities);
+    }
+
+    @Getter
+    @Setter
+    private String hubUrl = "https://hub.screener.io/wd/hub";
```
Comment (nadvolod) [inline java/src/test/java/com/saucelabs/saucebindings/SauceVisualSession.java:45]: Codacy found an issue: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/manual/SauceLabs/sauce_bindings/pullRequest?prid=6239731)

File: java/src/main/java/com/saucelabs/saucebindings/visual/SauceVisualOptions.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    private String viewportSize;
```
Comment (nadvolod) [inline java/src/main/java/com/saucelabs/saucebindings/visual/SauceVisualOptions.java:9]: Codacy found an issue: [Perhaps 'viewportSize' could be replaced by a local variable.](https://app.codacy.com/manual/SauceLabs/sauce_bindings/pullRequest?prid=6239731)

File: java/src/main/java/com/saucelabs/saucebindings/visual/SauceVisualOptions.java (hunk 7 above)
Comment (nadvolod) [inline java/src/main/java/com/saucelabs/saucebindings/visual/SauceVisualOptions.java:8]: Codacy found an issue: [Perhaps 'visualProjectName' could be replaced by a local variable.](https://app.codacy.com/manual/SauceLabs/sauce_bindings/pullRequest?prid=6239731)

File: java/src/test/java/com/saucelabs/saucebindings/integration/VisualE2ETest.java [hunk 11]
```diff
@@ -0,0 +1,56 @@
+package com.saucelabs.saucebindings.integration;
+
+import com.saucelabs.saucebindings.SauceVisualSession;
+import com.saucelabs.saucebindings.visual.SauceVisualOptions;
+import lombok.var;
```
Comment (nadvolod) [inline java/src/test/java/com/saucelabs/saucebindings/integration/VisualE2ETest.java:5]: Codacy found an issue: [Avoid unused imports such as 'lombok.var'](https://app.codacy.com/manual/SauceLabs/sauce_bindings/pullRequest?prid=6239731)

File: java/src/main/java/com/saucelabs/saucebindings/visual/SauceVisualOptions.java (hunk 10 above)
Comment (nadvolod) [inline java/src/main/java/com/saucelabs/saucebindings/visual/SauceVisualOptions.java:9]: Codacy found an issue: [Avoid unused private fields such as 'viewportSize'.](https://app.codacy.com/manual/SauceLabs/sauce_bindings/pullRequest?prid=6239731)

**Comment by nadvolod** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 8
           

Complexity increasing per file
==============================
- java/src/main/java/com/saucelabs/saucebindings/visual/SauceVisualOptions.java  4
- java/src/test/java/com/saucelabs/saucebindings/integration/VisualE2ETest.java  1
- java/src/test/java/com/saucelabs/saucebindings/SauceVisualSession.java  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/SauceLabs/sauce_bindings/pullRequest?prid=6239731&bid=20327768)

File: java/src/main/java/com/saucelabs/saucebindings/visual/SauceVisualOptions.java (hunk 10 above)
Comment (nadvolod) [inline java/src/main/java/com/saucelabs/saucebindings/visual/SauceVisualOptions.java:9]: Codacy found an issue: [Perhaps 'viewportSize' could be replaced by a local variable.](https://app.codacy.com/manual/SauceLabs/sauce_bindings/pullRequest?prid=6239731)

File: java/src/main/java/com/saucelabs/saucebindings/visual/SauceVisualOptions.java (hunk 7 above)
Comment (nadvolod) [inline java/src/main/java/com/saucelabs/saucebindings/visual/SauceVisualOptions.java:8]: Codacy found an issue: [Avoid unused private fields such as 'visualProjectName'.](https://app.codacy.com/manual/SauceLabs/sauce_bindings/pullRequest?prid=6239731)

File: java/src/main/java/com/saucelabs/saucebindings/visual/SauceVisualOptions.java (hunk 10 above)
Comment (nadvolod) [inline java/src/main/java/com/saucelabs/saucebindings/visual/SauceVisualOptions.java:9]: Codacy found an issue: [Avoid unused private fields such as 'viewportSize'.](https://app.codacy.com/manual/SauceLabs/sauce_bindings/pullRequest?prid=6239731)

File: java/src/main/java/com/saucelabs/saucebindings/visual/SauceVisualOptions.java (hunk 6 above)
Comment (nadvolod) [inline java/src/main/java/com/saucelabs/saucebindings/visual/SauceVisualOptions.java:7]: Codacy found an issue: [Perhaps 'visualCapabilities' could be replaced by a local variable.](https://app.codacy.com/manual/SauceLabs/sauce_bindings/pullRequest?prid=6239731)

File: java/src/test/java/com/saucelabs/saucebindings/SauceVisualSession.java (hunk 9 above)
Comment (nadvolod) [inline java/src/test/java/com/saucelabs/saucebindings/SauceVisualSession.java:45]: Codacy found an issue: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/manual/SauceLabs/sauce_bindings/pullRequest?prid=6239731)

File: java/src/test/java/com/saucelabs/saucebindings/SauceVisualSession.java (hunk 8 above)
Comment (nadvolod) [inline java/src/test/java/com/saucelabs/saucebindings/SauceVisualSession.java:6]: Codacy found an issue: [Avoid unused imports such as 'lombok.var'](https://app.codacy.com/manual/SauceLabs/sauce_bindings/pullRequest?prid=6239731)

File: java/src/main/java/com/saucelabs/saucebindings/visual/SauceVisualOptions.java (hunk 7 above)
Comment (nadvolod) [inline java/src/main/java/com/saucelabs/saucebindings/visual/SauceVisualOptions.java:8]: Codacy found an issue: [Perhaps 'visualProjectName' could be replaced by a local variable.](https://app.codacy.com/manual/SauceLabs/sauce_bindings/pullRequest?prid=6239731)

File: java/src/test/java/com/saucelabs/saucebindings/integration/VisualE2ETest.java (hunk 11 above)
Comment (nadvolod) [inline java/src/test/java/com/saucelabs/saucebindings/integration/VisualE2ETest.java:5]: Codacy found an issue: [Avoid unused imports such as 'lombok.var'](https://app.codacy.com/manual/SauceLabs/sauce_bindings/pullRequest?prid=6239731)

**Comment by titusfortner** [comment]
Should we have a separate session class for Visual?

We could do something as simple as:

```
SauceOptions options = new SauceOptions();
options.setVisual();

SauceSession session = new SauceSession(options);
```

We don't necessarily need a VisualOptions class either:
```
SauceOptions options = new SauceOptions();
options.setName("visualProjectName");
options.setScreenResolution("1280x1024") // same as viewportSize
```

We could even use a `takeScreenshot()` method that would work for normal Sauce sessions
```
session.takeScreenshot("Snapshot name");
```

**Comment by nadvolod** [comment]
@titusfortner I really like your approach. It's certainly cleaner. I went to separate classes because it seemed like that's the same direction we were going in mobile. So I tried to keep it similar. Regardless, I do want to iron out the use case below. To me, there is ambiguity when using/re-using methods because it means that they will have different behavior depending on whether you are using functional or visual, potentially breaking SRP.

```
    public void withVisualOptions() {
        SauceOptions options = new SauceOptions();
        //This becomes a mandatory call in order to start calling the methods below,
        //otherwise they become useless/confusing methods?
        options.setVisual();

        // is it clear that setProjectName only applies to Sauce Visual and not to Functional?
        options.setProjectName("Sauce Bindings");

        // how do we know that this is for visual?
        options.setViewportSize("1280x1024");

        var visualSession = new SauceSession(visualOptions);

        //Not sure if you always need to set test name,
        //but if we do then maybe we can just do start(testName);
        visualSession.start();

        visualSession.setTestName("Visual Test");

        //Same thing here, do we want ambiguity between visual and functional so that when we call this method,
        // users might be confused about why the snapshot isn't happening in their functional test
        visualSession.takeSnapshot("Snapshot name 1");

        Boolean isPassed = visualSession.stop();
        assertTrue(isPassed);
    }
```

This seems like a blend of both and solves the ambiguity:
```
    public void withVisualOptions() {
        SauceOptions options = new SauceOptions();
        options.setVisual();
        options.visual.setProjectName("Sauce Bindings");
        options.visual.setViewportSize("1280x1024");
```

**Comment by titusfortner** [comment]
Superceded by #256

**Pull request closed without merging**